src/native/s3select/s3select_napi.cpp

```diff
@@ -0,0 +1,289 @@
+    S3SelectNapi(const Napi::CallbackInfo& info);
+    Napi::Value Write(const Napi::CallbackInfo& info);
+    Napi::Value Flush(const Napi::CallbackInfo& info);
+    ~S3SelectNapi();
+
+private:
+    static Napi::FunctionReference constructor;
+    InputFormat input_format;
+    s3selectEngine::s3select s3select;
+    s3selectEngine::csv_object* csv_object = nullptr;
+    s3selectEngine::json_object* json_object = nullptr;
+    const uint8_t* headers_buf;
+    uint32_t headers_len;
+};
+
+Napi::Value
+S3SelectNapi::Write(const Napi::CallbackInfo& info)
+{
+    Napi::Buffer<char> buffer = info[0].As<Napi::Buffer<char>>();
+    SelectWorker* worker = new SelectWorker(
+        info,
+        *this,
+        csv_object,
+        json_object,
+        buffer.Data(),
+        buffer.Length(),
+        false,
+        input_format,
+        headers_buf,
+        headers_len);
+    worker->Queue();
+    return worker->_deferred.Promise();
+}
+
+Napi::Value
+S3SelectNapi::Flush(const Napi::CallbackInfo& info)
+{
+    SelectWorker* worker = new SelectWorker(
+        info,
+        *this,
+        csv_object,
+        json_object,
+        nullptr, /*No buffer for flush*/
+        0,
+        true,
+        input_format,
+        headers_buf,
+        headers_len);
+    worker->Queue();
+    return worker->_deferred.Promise();
+}
+
+Napi::FunctionReference S3SelectNapi::constructor;
+
+Napi::Object
+S3SelectNapi::Init(Napi::Env env, Napi::Object exports)
+{
+    Napi::HandleScope scope(env);
+
+    Napi::Function func = DefineClass(
+        env,
+        "S3SelectNapi",
+        {
+            InstanceMethod("write", &S3SelectNapi::Write),
+            InstanceMethod("flush", &S3SelectNapi::Flush)
+        }
+    ); //end of DefineClass
+
+    constructor = Napi::Persistent(func);
+    constructor.SuppressDestruct();
+
+    exports.Set("S3Select", func);
+    return exports;
+}
+
+void
+s3select_napi(Napi::Env env, Napi::Object exports)
+{
+    S3SelectNapi::Init(env, exports);
+}
+
+std::string
+GetStringWithDefault(Napi::Object obj, std::string name, std::string dfault)
+{
+    if (obj.Has(name)) {
+        return obj.Get(name).ToString().Utf8Value();
+    }
+    return dfault;
+}
+
+S3SelectNapi::S3SelectNapi(const Napi::CallbackInfo& info)
+    : Napi::ObjectWrap<S3SelectNapi>(info)
+{
+    Napi::Object context = info[0].As<Napi::Object>();
+    Napi::Object input_serialization_format = context.Get("input_serialization_format").As<Napi::Object>();
+    Napi::Buffer headers_buf_obj = context.Get("records_header_buf").As<Napi::Buffer<uint8_t>>();
+    headers_buf = headers_buf_obj.Data();
+    headers_len = headers_buf_obj.Length();
+    std::string query = context.Get("query").ToString();
+    std::string input_format_str = context.Get("input_format").ToString();
+
+    if (!input_format_str.compare("JSON")) {
+        input_format = InputFormat::json;
+    } else if (!input_format_str.compare("CSV")) {
+        input_format = InputFormat::csv;
```

Comment:
Comparing std::string is better with simple `==`, no need to use the compare method.
Also - why do we need to define another enum? I would simply send the string to the worker, and just define couple of consts to be used for checking equality - 
```c++
const char *CSV_FORMAT = "CSV";
const char *JSON_FORMAT = "JSON";
const char *PARQUET_FORMAT = "PARQUET";

...
if (input_format == CSV_FORMAT) {
  ...
} else if (input_format == JSON_FORMAT) {
  ...
}
```